pkg/controller/deployment/deployment_controller.go

```diff
@@ -599,6 +605,24 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
 	return true, err
 }
 
+func (dc *DeploymentController) cleanupOldRcs(controllers []*api.ReplicationController, deployment extensions.Deployment) error {
+	diff := len(controllers) - *deployment.Spec.RevisionHistoryLimit
+	if diff <= 0 {
+		return nil
+	}
+
+	sort.Sort(byCreationTimestamp(controllers))
+	// TODO: This should be parallelized.
+	for i := 0; i < diff; i++ {
+		controller := controllers[i]
+		if err := dc.client.ReplicationControllers(controller.Namespace).Delete(controller.Name); err != nil {
```

Comment:
Since we avoid delete RCs with non-zero replica counts, maybe we need add a gc routine to delete those leftover rcs periodically.